### DISCUSSION TITLE
How do I understand the component, and how to set its value?

### DISCUSSION THREAD

**Discussion by squarefaceyao** [lululxvi/deepxde#859] [Q&A] [answered]
Dear Lu Lu,
Thanks for the great support for DeepXDE, and it's powerful. 

How do I understand the component, and how to set its value? Take the following code as an example. 

```
dde.icbc.IC(geom, lambda X: 0.05, boundary, component=0)
dde.icbc.PointSetBC(observe_t, ob_y[:, 1:2], component=0)
```

**Comment by lululxvi** [comment, marked as answer]
See
- https://deepxde.readthedocs.io/en/latest/modules/deepxde.icbc.html#deepxde.icbc.boundary_conditions.BC
- https://deepxde.readthedocs.io/en/latest/modules/deepxde.icbc.html#deepxde.icbc.boundary_conditions.PointSetBC

**Comment by squarefaceyao** [comment]
Thank you very much for your help.
I don't understand what's in the link. I really appreciate your help. 
If my data is one-dimensional, is the value of component set to 0?


### My problem is described as follows
We will solve a simple ODE system:

$$  {\frac{dV}{dt}}=10- {G_{Na}m^3h(V-50)} - {G_{K}n^4(V+77)} - {G_{L}(V+54.387)} $$ 

$$  {\frac{dm}{dt}}=\left(\frac{0.1{(V+40)}}{1-e^\frac{-V-40}{10}}\right)(1-m) - \left(4e^{\frac{-V-65}{18}}\right)m $$

$$  \frac{dh}{dt}= {\left(0.07e^{\frac{-V-65}{20}}\right)(1-h)} - \left(\frac{1}{1+e^\frac{-V-35}{10}}\right)h $$

$$  \frac{dn}{dt}= {\left(\frac{0.01(V+55)}{1-e^\frac{-V-55}{10}}\right)}(1-n) - \left(0.125e^{\frac{-V-65}{80}}\right)n $$

$$ \qquad \text{where} \quad t \in [0,7], $$

with the initial conditions  

$$ V(0) = -65,  m(0) = 0.05 ,  h(0) = 0.6 ,  n(0) = 0.32 $$

The reference solution is [here](https://drive.google.com/file/d/1gtq0Gwi06170MOAGoxOLPGT5pJ54Jb2d/view?usp=sharing), where the parameters $G_{na},G_{k},G_{L}$ are gated variables and whose true values are 120, 36, and 0.3, respectivly. 

## My code can't predict the parameters correctly. I would like to ask where is the problem.
```python
import deepxde as dde
import numpy as np
import torch
C1 = dde.Variable(50.0) # g_Na
C2 = dde.Variable(50.0) # g_K
C3 = dde.Variable(1.0) # g_L

# Most backends
def HH(t, y):
    
    
    V, m, h, n= y[:, 0:1], y[:, 1:2], y[:, 2:3], y[:, 3:4]
    dy1_x = dde.grad.jacobian(y, t, i=0)
    dy2_x = dde.grad.jacobian(y, t, i=1)
    dy3_x = dde.grad.jacobian(y, t, i=2)
    dy4_x = dde.grad.jacobian(y, t, i=3)
    
    def I_inj(t):
        return 10*(t>10) - 10*(t>20) + 35*(t>30) - 35*(t>40)
        
    def alpha_m(V):
        return 0.1*(V+40.0)/(1.0 - torch.exp(-(V+40.0) / 10.0))
        
    def beta_m(V):
        return 4.0*torch.exp(-(V+65.0) / 18.0)
    
    def alpha_h(V):
        return 0.07*torch.exp(-(V+65.0) / 20.0)
    
    def beta_h(V):
        return 1.0/(1.0 + torch.exp(-(V+35.0) / 10.0))
    
    def alpha_n(V):
        return 0.01*(V+55.0)/(1.0 - torch.exp(-(V+55.0) / 10.0))
    
    def beta_n(V):
        return 0.125*torch.exp(-(V+65) / 80.0)
        
    
    return [
        dy1_x - I_inj(t) - C1 * m**3 * h * (V - 50) - C2 * n**4 * (V + 77.0) - C3 * (V + 54.387),
        dy2_x - alpha_m(V)*(1.0-m) - beta_m(V)*m,
        dy3_x - alpha_h(V)*(1.0-h) - beta_h(V)*h,
        dy4_x - alpha_n(V)*(1.0-n) - beta_n(V)*n,
    ]


def boundary(_, on_initial):
    return on_initial


geom = dde.geometry.TimeDomain(0, 20)

# Initial conditions -65, 0.05, 0.6, 0.32
ic1 = dde.icbc.IC(geom, lambda X: -65, boundary, component=0) # 数据的初始数值
ic2 = dde.icbc.IC(geom, lambda X: .05, boundary, component=0)
ic3 = dde.icbc.IC(geom, lambda X: .6, boundary, component=0)
ic4 = dde.icbc.IC(geom, lambda X: .32, boundary, component=0)


# 使用仿真的数据 the train data
observe_t = np.expand_dims(observe_t,axis=1) # 扩充纬度
observe_y0 = dde.icbc.PointSetBC(observe_t, ob_y[:, 0:1], component=0)
observe_y1 = dde.icbc.PointSetBC(observe_t, ob_y[:, 1:2], component=0)
observe_y2 = dde.icbc.PointSetBC(observe_t, ob_y[:, 2:3], component=0)
observe_y3 = dde.icbc.PointSetBC(observe_t, ob_y[:, 3:4], component=0)


data = dde.data.PDE(
    geom,
    HH,
    [ic1, ic2,ic3,ic4, observe_y0, observe_y1, observe_y2, observe_y3],
    num_domain=4000,
    num_boundary=6,
    anchors=observe_t,
)

net = dde.nn.FNN([1] + [32] * 3 + [4], "sin", "Glorot uniform")
model = dde.Model(data, net)
epochs = 60000
vari_save_filename="fhn_variables_datasize_{}_epochs_{}.dat".format(ob_y.shape[0],epochs)

model.compile("adam", lr=0.00001, external_trainable_variables=[C1, C2, C3])
variable = dde.callbacks.VariableValue(
    [C1, C2, C3], filename=vari_save_filename
)
losshistory, train_state = model.train(iterations=epochs, callbacks=[variable])

print(variable.get_value())
# dde.saveplot(losshistory, train_state, issave=True, isplot=True)
```
### My complete code is in this link https://gitee.com/squarefaceyao/pinn_inverse_pes/blob/master/Simulation_HH.ipynb

**Comment by lululxvi** [comment]
See the ODE system example at https://deepxde.readthedocs.io/en/latest/demos/pinn_inverse.html
